design/KEPS/namespace-discovery/proposal.md

```diff
@@ -99,6 +99,30 @@ TODO: external controlplane
 
 Since the discovery selectors act as a way to separate multiple Kiali deployments from one another, when Kiali relies on autodiscovering the Istio discovery selectors and it cannot auto discover them, the Kiali server will fail to start and/or panic rather than potentially show out of scope data. In a multi-primary deployment, if one primary cannot be reached then that cluster will be considered inaccessible.
 
+## Discovery Selector Kiali Config Option
+
+A new configuration option will be added that will mimic the Istio discovery selectors configuration option. Using the configuration option, users will be able to specify which namespaces Kiali should be scoped to. Having a config option within Kiali has the advantage of not relying on autodiscovery of the Istio controlplanes. In cases where Kiali temporarily cannot access the control plane configuration, Kiali does not need to fail since it will rely on its own discovery selectors to scope itself. The configuration option needs to provide enough flexibility for users to specify different discovery selectors for each cluster in multi-primary scenarios.
+
+Kiali would add a top level `discoverySelectors` config option where you can specify a `map[string]labelselector` where the key is the cluster name that the discovery selector will apply to.
```

Comment:
I think perhaps we should have a global discoverySelector and only cluster overrides.  This would allow a simple definition that can apply to all clusters, which makes things easy for single-cluster and scenarios where all clusters use the same thing.

Reply:
I like that option.

Reply:
Is it a valid edge case that clusters can have different DS? I'm sure Istio allows it (because, one control plane probably doesn't care about another control plane's DS) but is it a valid thing to do (do people do that?). 

And I thought Kiali assumes "namespace sameness" in multi-cluster?

Reply:
> Is it a valid edge case that clusters can have different DS? I'm sure Istio allows it (because, one control plane probably doesn't care about another control plane's DS) but is it a valid thing to do (do people do that?).

I'd say yes. It's perfectly valid to have namespaces that exist in one cluster but don't exist on another.

> And I thought Kiali assumes "namespace sameness" in multi-cluster?

You can't assume that the same namespace will exist on both clusters and that the two clusters will be exactly the same. The actual namespace object only exists inside a single cluster even if istio logically treats an app deployed in the same namespace across both clusters as the same app.

Reply:
> We could probably hold off on the cluster-specific impl until it's proven we need it.

I'd rather give users the ability to specify this now or at the very least keep the structure the same where we have a `global:` setting so that we can easily add cluster specific overrides later.

Reply:
I'll add "global:" to all the examples where there is none here in this document. The one issue I'm wrestling with is how will the operator know the home cluster name so it can do its own override when processing the selectors (because it needs to do this in order to create the Roles in the proper places).

For example, suppose we have this, for a hard multi-tenancy setup:

```yaml
spec:
  deployment:
    cluster_wide_access: false
    discovery_selectors:
      global:
      - match_labels:
          foo: howdy
      cluster-a:
      - match_labels:
          foo: hello
      cluster-b:
      - match_labels:
          foo: wassup
```

Suppose Kiali's home cluster is `cluster-b` -- how does the operator know that? Because the operator will need to know that, since it will need to know that `cluster-b` is overriding `global`.

(BTW: I'm still not clear on what it means for one cluster's selectors to override the global selectors - I assume it means the cluster-specific one just completely overrides the global, where I could have a dozen more clusters, but because they aren't specified in `discovery_selectors`, then just the `global` ones are used, but for cluster-a and cluster-b, the entire array of selectors under global are ignored.)

Reply:
Hmm.. I think I might have an answer to my question above about "how does the operator know the home cluster?". The operator can look at [spec.clustering.clusters[*].name](https://kiali.io/docs/configuration/kialis.kiali.io/#.spec.clustering.clusters[*].name) - those are all the remote cluster names. If there is an override cluster section under "discovery_selectors" that does NOT appear in that list of remote cluster names, it must be the home cluster. This assumes every remote cluster is specified in that remote cluster section (even those not accessible).

Reply:
Instead of `global` maybe it should be `default`.  I would expect if there is a cluster-override it would not be additive, it would be a replacement.

Reply:
We could also say the "global" (or "default") simply IS the home cluster. So maybe instead of global or default we call it "home" or "kiali-home" or "local" (we want it unique enough for it not to conflict with any cluster name a user might want to use - so "kiali-home" seems like something people won't use, but "local" and "home" seem more useful cluster names for users to want to use).

So you would define the kiali home cluster and it is used as the default (which, when you think about it, is likely going to be the common global selectors anyway - at least for the majority of use cases. So that seems like a viable "default" for "global" selectors).

So instead of:

```yaml
deployment_selectors:
  global:
    foo: hello
  cluster-a:
    foo: howdy
```

it would be:
```yaml
deployment_selectors:
  kiali-home:
    foo: hello
  cluster-a:
    foo: howdy
```

That seems to make sense to me

Reply:
I also think that we should change "global" for no other reason than that seems like it could potentially be a cluster name that a user might want ("global" as in "worldwide", then you have different regions like "emea", "na", etc).

I think it should be "kiali-home" because (a) it is more descriptive and (b) likely won't clash with any name a user might want to use.

Reply:
we ended up just going with "default" for now.

Reply:
> The one issue I'm wrestling with is how will the operator know the home cluster name so it can do its own override when processing the selectors (because it needs to do this in order to create the Roles in the proper places)

The operator should not be cluster aware. Today iirc the "hard multi-tenancy mode" of `accessible_namespaces` does not work for multi-cluster. Unless that's changing the operator would just configure the roles for the cluster that the operator is deployed on and not try to do anything cluster related.

Reply:
> I also think that we should change "global" for no other reason than that seems like it could potentially be a cluster name that a user might want ("global" as in "worldwide", then you have different regions like "emea", "na", etc).

Hmmm that's a good point and any name will eventually conflict so imo we should just change the structure to something like:
```
deployment_selectors:
  global (or default):
    foo: hello
  overrides:
    <cluster-name>:
      foo: howdy
```
Merging global selectors with override selectors would be more complicated so I think the simplest heuristic would be to only use what is specified in `overrides` if you have specified anything.

Reply:
I have no problem with adding an `overrides`.  We already preferred `default`, and `overrides` works well with `default`.   I'm -1, though on ignoring `default` if

Reply:
> I think you mean that if there is a cluster-specific override that you would ignore default for that cluster, and any cluster not included in overrides would get the defaults.

Yes exactly

Reply:
OK. let's change this KEP to refer to this extra level "overrides".